src/pytest_plugins/logging.py

```diff
@@ -0,0 +1,262 @@
+        """Apply colorful formatting."""
+        if self.running_in_docker():
+            return super().format(record)
+        color = self.COLORS.get(record.levelno, self.RESET)
+        record.levelname = f"{color}{record.levelname}{self.RESET}"
+        return super().format(record)
+
+    @staticmethod
+    def running_in_docker() -> bool:
+        """Return True if `/.dockerenv` exists."""
+        return Path("/.dockerenv").exists()
+
+
+class LogLevel:
+    """Help parse a log-level provided on the command-line."""
+
+    @classmethod
+    def from_cli(cls, value: str) -> int:
+        """
+        Parse a logging level from CLI.
+
+        Accepts standard level names (e.g. 'INFO', 'debug') or numeric values.
+        """
+        try:
+            return int(value)
+        except ValueError:
+            pass
+
+        level_name = value.upper()
+        if level_name in logging._nameToLevel:
+            return logging._nameToLevel[level_name]
+
+        valid = ", ".join(logging._nameToLevel.keys())
+        raise ValueError(f"Invalid log level '{value}'. Expected one of: {valid} or a number.")
+
+
+def pytest_addoption(parser):  # noqa: D103
+    logging_group = parser.getgroup(
+        "logging", "Arguments related to logging from test fixtures and tests."
+    )
+    logging_group.addoption(
+        "--eest-log-level",  # --log-level is defined by pytest's built-in logging
+        "--eestloglevel",
+        action="store",
+        default="INFO",
+        type=LogLevel.from_cli,
+        dest="eest_log_level",
+        help=(
+            "The logging level to use in the test session: DEBUG, INFO, WARNING, ERROR or "
+            "CRITICAL, default - INFO. An integer in [0, 50] may be also provided."
+        ),
+    )
+
+
+@functools.cache
+def get_log_stem(argv0: str, argv1: Optional[str]) -> str:
+    """Generate the stem (prefix-subcommand-timestamp) for log files."""
+    stem = Path(argv0).stem
+    prefix = "pytest" if stem in ("", "-c", "__main__") else stem
+    subcommand = argv1 if argv1 and not argv1.startswith("-") else None
+    timestamp = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
+
+    name_parts = [prefix]
+    if subcommand:
+        name_parts.append(subcommand)
+    name_parts.append(timestamp)
+
+    return "-".join(name_parts)
+
+
+def pytest_configure_node(node):
+    """Initialize a variable for use in the worker (xdist hook)."""
+    potential_subcommand = None
+    if len(sys.argv) > 1:
+        potential_subcommand = sys.argv[1]
+    node.workerinput["log_stem"] = get_log_stem(sys.argv[0], potential_subcommand)
+
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_configure(config: pytest.Config) -> None:
+    """
+    Initialize logging.
+
+    This goes to a lot of effort to ensure that a log file is created per worker
+    if xdist is used and that the timestamp used in the filename is the same across
+    main and all workers.
+    """
+    global file_handler
```

Comment:
I don't quite understand why this line is necessary but I commented out and nothing seemed to change, even with xdist enabled to two processes.

Reply:
I also don't have a lot of experience with the `global` keyword. If this line is removed, then no "`file_handler` only" logs are written to the `main` log file with xdist > 1. They are written however if it's removed, without xdist. It probably fails to work with xdist>1 because the workers don't share the file handler with the main thread; they just have a different, locally defined `file_handler`. But I'm not entirely sure tbh.